src/server/completions/completions.ts

```diff
@@ -46,16 +64,275 @@ export function getCompletionItems(
+  else {
+    return [];
+  }
+}
+
+function bucketMatchingFieldNames(fields: Field[], fieldToMatch: string) {
+  const queries: string[] = [];
+  const dimensions: string[] = [];
+  const measures: string[] = [];
+  const explores: string[] = [];
+  for (const field of fields) {
+    if (fieldToMatch.length === 0 || field.name.startsWith(fieldToMatch)) {
+      const type = fieldType(field);
+      if (isFieldAggregate(field)) {
+        measures.push(field.name);
+      } else if (field.isExploreField()) {
+        explores.push(field.name);
+      } else if (type === 'query') {
+        queries.push(field.name);
+      } else {
+        dimensions.push(field.name);
+      }
+    }
+  }
+  return {queries, dimensions, measures, explores};
+}
+
+export interface DocumentTextParse {
+  /**
+   * Document text containing `Explore`s except for anonymous query-level source definitions
+   *  - includes all unaltered imports
+   *  - includes all unaltered source definitions
+   */
+  truncatedText: string;
+  /**
+   * Number of top-level imports and source definitions
+   */
+  exploreCount: number;
+  /**
+   * Document text for only the top-level definition that the cursor currently resides in
+   * or nearest to, undefined if cursor is definitely not within a top-level definition
+   */
+  nearestDefinitionText?: string[];
+  /**
+   * The adjusted cursor position within the nearest top-level definition text as stored in
+   * `nearestSymbolText`, undefined if cursor is not inside a top-level definition
+   */
+  adjustedCursor?: Position;
+}
+
+function parseDocumentText(text: string, cursor: Position): DocumentTextParse {
+  const parse: DocumentTextParse = {
+    truncatedText: '',
+    exploreCount: 0,
+  };
+  const truncatedLines: string[] = [];
+  const topLevelSymbols = /\b((?:source|query|run|sql)\s*:|import)\s*/gi;
+
+  let lastSymbolStart: Position | undefined = undefined;
+  let lastNonemptyLine = -1;
+  let includeLines = false;
+
+  const lines = text.split('\n');
+  for (const [i, line] of lines.entries()) {
+    const symbolMatches = topLevelSymbols.exec(line);
+    if (symbolMatches != null && symbolMatches.index === 0) {
+      if (
+        symbolMatches[1].toLowerCase().startsWith('source') ||
+        symbolMatches[1].toLowerCase().startsWith('import')
+      ) {
+        includeLines = true;
+        parse.exploreCount++;
+      } else {
+        includeLines = false;
+      }
+      parseCurrentDefinition(
+        lastSymbolStart,
+        cursor,
+        lastNonemptyLine,
+        parse,
+        lines
+      );
+      lastSymbolStart = {
+        line: i,
+        character: symbolMatches.index,
+      };
+    }
+    if (includeLines) {
+      truncatedLines.push(line);
+    }
+    if (!isEmptyLine(line)) {
+      lastNonemptyLine = i;
+    }
+  }
+  parseCurrentDefinition(
+    lastSymbolStart,
+    cursor,
+    lastNonemptyLine,
+    parse,
+    lines
+  );
+  parse.truncatedText = truncatedLines.join('\n');
```

Comment:
One thing that is done with the malloy-sql parsing is that instead of omitting lines, blank lines are used so that line offset math doesn't have to happen.

Reply:
To avoid a refactor given time constraints, I've elected to leave this since cursor math is minimal